### PR TITLE
Write snippets index file when generating docs

### PR DESCRIPTION
**Pull request by DanTup**
I don't know if this is a good/the best way to do this, but I figured it was worth opening as a starting point. We'd like to be able to present users with a list of the samples that are available for creation which means having a list somewhere. This writes a `snippets/txt` file alongside the `snippets` folder that just contains a list of the snippets IDs (filenames minus the `.dart` extension).

@jcollins-g @gspencergoog Is this reasonable? (If not, please point me in the right direction :-))

And if so, do we normally have tests for this code like this - is there somewhere I should add something?

(cc @pq)

### CODE REVIEW

**Comment by gspencergoog** [comment]
The Dartdoc tool doesn't have tests, but the snippets code does:
https://github.com/flutter/flutter/tree/master/dev/snippets/test

**Comment by DanTup** [comment]
> The easiest way to do this would be to have the Snippets tool write out a second file for each snippet that has metadata in it about the file (source file path, line number range, etc)

I thought about this too - I wasn't sure if there was other metadata we'd need in addition to the name. If that other info is useful, I don't mind taking a look at that - writing the JSON file shouldn't be difficult, and then it's just tweaking this code to read and combine the JSON rather than the `.dart` files (assuming where this code currently runs is the correct place to combine the JSON?).

**Comment by gspencergoog** [comment]
The correct place to write the JSON is in the snippets tool, since it has the metadata, and the right place to combine it is in the dartdoc.dart file: they can all live in the snippets output directory, and firebase will automatically upload them all. It seems useful to have the metadata available anyhow, both as a combined index, and as individual files: the "flutter create" command could be printing out more information about the context of the snippet, for instance.

You'd probably write it out [here](https://github.com/flutter/flutter/blob/cbb168e7b9f40e77d595f65f9940526dc32bc044/dev/snippets/lib/main.dart#L145), since that's where you have access to the metadata (it's in the parsed args). You can use the configuration object from the `SnippetGenerator` object to find where to write the output.

**Comment by DanTup** [comment]
> write out a second file for each snippet that has metadata in it about the file (source file path, line number range, etc)

What metadata do we need, and where is it available? I'm looking in SnippetGenerator but I'm not surewhat line number range we need (for ex).

**Comment by DanTup** [comment]
Ok, I pushed changes to write a JSON metadata file and combine them into an array, so the output looks like this:

```text
dantup-macbookpro:flutter dantup$ ls dev/docs/doc/snippets/
app_bar.AppBar.actions.dart                           material.Card.dart
app_bar.AppBar.actions.dart.json                      material.Card.dart.json
card.Card.dart                                        material.DeletableChipAttributes.onDeleted.dart
card.Card.dart.json                                   material.DeletableChipAttributes.onDeleted.dart.json
chip.DeletableChipAttributes.onDeleted.dart           material.IconButton.dart
chip.DeletableChipAttributes.onDeleted.dart.json      material.IconButton.dart.json
icon_button.IconButton.dart                           material.Scaffold.dart
icon_button.IconButton.dart.json                      material.Scaffold.dart.json
index.json                                            scaffold.Scaffold.dart
material.AppBar.actions.dart                          scaffold.Scaffold.dart.json
material.AppBar.actions.dart.json
```

```js
dantup-macbookpro:flutter dantup$ cat dev/docs/doc/snippets/index.json 
[
{
    "file": "card.Card.dart"
},
{
    "file": "app_bar.AppBar.actions.dart"
},
{
    "file": "chip.DeletableChipAttributes.onDeleted.dart"
},
{
    "file": "material.Scaffold.dart"
},
{
    "file": "material.AppBar.actions.dart"
},
{
    "file": "material.DeletableChipAttributes.onDeleted.dart"
},
{
    "file": "material.Card.dart"
},
{
    "file": "icon_button.IconButton.dart"
},
{
    "file": "material.IconButton.dart"
},
{
    "file": "scaffold.Scaffold.dart"
}
]
```

We just need to refine what goes into the JSON (and add some tests).

File: dev/snippets/lib/snippets.dart
Comment (gspencergoog) [inline dev/snippets/lib/snippets.dart]: I was thinking that you probably actually want to write it in the main.dart, since this class doesn't have all of the metadata available. But, if you want to write it here (which makes sense, other than the metadata not being available), you could pass in a Map with the data in it from the main.

File: dev/snippets/lib/snippets.dart
Comment (gspencergoog) [inline dev/snippets/lib/snippets.dart]: I would prefer the name to be just ".json", not ".dart.json".

File: dev/tools/dartdoc.dart
Comment (gspencergoog) [inline dev/tools/dartdoc.dart]: It might be better to actually interpret the JSON, add the object to an array, and write it out with the JSON encoder. Then things like indentation could be correct, and it would be easier to switch the representation to something else or add index-wide data (creation date of the index, etc.) if we wanted to. This isn't performance critical code.

**Review by gspencergoog — CHANGES_REQUESTED** [review]

**Comment by jcollins-g** [comment]
This all looks pretty reasonable to me, and given that you want to make the canonical location for this data served via the API docs, writing out the index file here is OK too.  I was kind of assuming that for IDE usage (especially offline) the file would be better placed as a generated file in the repository -- if so, I'd imagine you'd want this in flutter_tools somewhere.

**Comment by pq** [comment]
> What metadata do we need

I think minimally we want a presentable label.  Possibly a description too.  For example,  looking at https://docs.flutter.io/flutter/material/Card-class.html, it would be great to have access to this description in the metadata. 

    "This sample shows creation of a Card widget that shows album information and two actions."

File: dev/snippets/lib/snippets.dart
Comment (DanTup) [reply dev/snippets/lib/snippets.dart]: I don't mind moving it - I couldn't see what metadata we wanted so didn't know where it existed. I'll move it to `main.dart` and add the metadata I can see (package, library, element names), but if there should be a range in there too, let me know where I can find (or calculate) it.

Comment (DanTup) [reply dev/snippets/lib/snippets.dart]: Actually, I'll leave it inside for now simple because the filename may be passed in as null and generated itself, and I'd rather not duplicate that logic outside to keep the filenames the same (should it ever change), and I'll pass in a map (though lmk if you'd prefer the opposite).

**Comment by DanTup** [comment]
Thanks all - I've pushed some tweaks that:

1. Add additional metadata to file (package, library, elementName, ID from `main.dart` and description from within `snippets.dart`)
2. Write a .json file isntead of .dart.json
3. Decode and re-encode JSON array for better formatting
4. Add a test that the metadata file is written

@pq note that some descriptions are longer than others and may contain newlines, so depending on where you show them they may need some tweaking (I didn't want to stop the newlines, as they may be paragraphs, and it's probably better for the IDE to shorten them than do it here, in case someone wants to show the full description).

Here's a sample of what the (combined) metadata file in `snippets/index.json` looks like when I run locally:

```js
[
    {
        "package": "flutter",
        "library": "material",
        "element": "Scaffold",
        "id": "material.Scaffold",
        "file": "material.Scaffold.dart",
        "description": "This example shows a [Scaffold] with an [AppBar], a [BottomAppBar] and a\n[FloatingActionButton]. The [body] is a [Text] placed in a [Center] in order\nto center the text within the [Scaffold] and the [FloatingActionButton] is\ncentered and docked within the [BottomAppBar] using\n[FloatingActionButtonLocation.centerDocked]. The [FloatingActionButton] is\nconnected to a callback that increments a counter."
    },
    {
        "package": "flutter",
        "library": "scaffold",
        "element": "Scaffold",
        "id": "scaffold.Scaffold",
        "file": "scaffold.Scaffold.dart",
        "description": "This example shows a [Scaffold] with an [AppBar], a [BottomAppBar] and a\n[FloatingActionButton]. The [body] is a [Text] placed in a [Center] in order\nto center the text within the [Scaffold] and the [FloatingActionButton] is\ncentered and docked within the [BottomAppBar] using\n[FloatingActionButtonLocation.centerDocked]. The [FloatingActionButton] is\nconnected to a callback that increments a counter."
    },
    {
        "package": "flutter",
        "library": "material",
        "element": "DeletableChipAttributes.onDeleted",
        "id": "material.DeletableChipAttributes.onDeleted",
        "file": "material.DeletableChipAttributes.onDeleted.dart",
        "description": "This sample shows how to use [onDeleted] to remove an entry when the\ndelete button is tapped."
    },
    // snip
```

@pq @gspencergoog btw, now I'm thinking maybe `flutter_tools` should consume this index file with a way to list the samples, rather than the editors using it directly? There was some logic in flutter tools to use either stable or master hostnames for `flutter create --sample` but if editors always pull this file from stable, they won't match up. WDYT?

**Comment by jcollins-g** [comment]
@DanTup @pq @gspencergoog Mostly an observer here, but I'll mention more strongly since you brought it up:  The most strictly correct path would be to have the snippets data generated somehow local to the flutter repository, rather than having a canonical source online.  That works better for offline usage, fixes problems of synchronicity, etc.

**Comment by pq** [comment]
> The most strictly correct path would be to have the snippets data generated somehow local to the flutter repository, rather than having a canonical source online. That works better for offline usage, fixes problems of synchronicity, etc.

Thanks for emphasizing this @jcollins-g.  I completely agree and worry, especially as the sample collection evolves, about version skew (e.g., might `flutter create` generate content that is not compatible with my actually installed SDK?)

**Comment by DanTup** [comment]
> The most strictly correct path would be to have the snippets data generated somehow local to the flutter repository, rather than having a canonical source online

That makes sense to me, though the `flutter create` functionality is already pulling live from the docs sites (and they need to be in sync).

@gspencergoog what do you think about changing `flutter create` to use locally stored snippets and changing where this writes? (and do we have an existing process for generating files in the flutter repo periodically?)

**Comment by pq** [comment]
@DanTup: instead of the name of the generated file:

```
{
    "file": "material.DeletableChipAttributes.onDeleted.dart"
}
```

(or in addition to) I wonder if we could add the path of the file that is hosting the snippet?  This would be useful, for example, when we want to produce banners on files w/ samples (https://github.com/flutter/flutter-intellij/issues/2976).

**Comment by DanTup** [comment]
@pq 

> I wonder if we could add the path of the file that is hosting the snippet?

I've added this as `sourcePath`, so now looks like this:

```js
[
    {
        "sourcePath": "lib/src/material/scaffold.dart",
        "package": "flutter",
        "library": "material",
        "element": "Scaffold",
        "id": "material.Scaffold",
        "file": "material.Scaffold.dart",
        "description": "This example shows a [Scaffold] with an [AppBar], a [BottomAppBar] and a\n[FloatingActionButton]. The [body] is a [Text] placed in a [Center] in order\nto center the text within the [Scaffold] and the [FloatingActionButton] is\ncentered and docked within the [BottomAppBar] using\n[FloatingActionButtonLocation.centerDocked]. The [FloatingActionButton] is\nconnected to a callback that increments a counter."
    },
    {
        "sourcePath": "lib/src/material/scaffold.dart",
        "package": "flutter",
        "library": "scaffold",
        "element": "Scaffold",
        "id": "scaffold.Scaffold",
        "file": "scaffold.Scaffold.dart",
        "description": "This example shows a [Scaffold] with an [AppBar], a [BottomAppBar] and a\n[FloatingActionButton]. The [body] is a [Text] placed in a [Center] in order\nto center the text within the [Scaffold] and the [FloatingActionButton] is\ncentered and docked within the [BottomAppBar] using\n[FloatingActionButtonLocation.centerDocked]. The [FloatingActionButton] is\nconnected to a callback that increments a counter."
    },
// ...
```

This still needs tests, but I'll wait for a decision on whether we should be reading these locally (@gspencergoog - this would mean changing the existing code in `flutter create` to not pull from the web) since the code might end up in a different place if that happens.

**Comment by pq** [comment]
Thanks @DanTup!

**Review by gspencergoog — APPROVED** [review]

**Comment by DanTup** [comment]
@gspencergoog I take it you're happy for this to remain "online" for now? I need to add some tests before merging, but wanted to be sure we're happy with this approach since there is discussion about about moving it offline (eg. bundling inside Flutter), though that would require additional changes to `flutter create`.

**Comment by gspencergoog** [comment]
Yes, I'm happy with it being online for now.

**Comment by DanTup** [comment]
Oops, turns out I'd already added a test. I've pushed a tiny addition to it to ensure the metadata passed into the generator (eg. `sourcePath`) is also in the output JSON though.

I'll land this on green then, thanks!

**Comment by gspencergoog** [comment]
Actually don't use that address, use this one instead: <https://master-docs.flutter.io/snippets/index.json>

Using the other one makes it harder to move the firebase hosting around.

**Comment by gspencergoog** [comment]
@DanTup, what happens when there are multiple instances in one file? Shouldn't the metadata also include the source line that it found the snippet at, in order to disambiguate them?

**Comment by DanTup** [comment]
Yeah, that's a fair point. I was thinking about just for listing them all for creation, but if we want to match them up to the code the user is looking at, the range information would be useful.

However, the possibility of being out of sync becomes a much bigger issue here, as the editor could give the samples in completely the wrong code (and not give them where it should) 😞

**Comment by DanTup** [comment]
(also, is the range information already available somewhere? I did look when you first mentioned it, but couldn't find anything obvious)

**Comment by DanTup** [comment]
Oh yeah, of course - that's where I got SOURCE_PATH! =D

PR here:

https://github.com/flutter/flutter/pull/26274

This doesn't solve the issue mentioned above about them being out of sync, but at least the info is available for the common case. Another thought I had was comparing `elementName` to the Outline data we have - I haven't tried implementing this yet though (maybe @pq knows whether this matches up well enough).

**Pull request merged**
